### ISSUE TITLE
Memory layouts for string processing

### ISSUE THREAD

**Issue by wesm** [weld-project/weld#136]
hi folks,

I'm excited about the Weld project. I have been looking at the Weld data structures and way that the runtime interacts with memory and have some questions, particularly about non-numeric data. 

I see here https://github.com/weld-project/weld/blob/master/python/grizzly/numpy_weld_convertor.cpp#L155 that a Weld string vector is semantically a vector of pointers. While this is one possible way to deal with arrays of variable-length types, I am wondering what it would take to expand to other kinds of non-pointer-based memory layouts, which can yield better processing efficiency for the CPU.

In pandas for example, our likely long term plan is to move toward a "packed" columnar memory model (as specified in Apache Arrow) for strings that is like:

```
length: 3
validity_bits [0 0 0 0 0 1 1 1] + padding for alignment
offsets: [0, 3, 6, 9] + padding
data: 'foobarbazqux' + padding
```

Beyond "packing" the strings in a contiguous buffer, you can also dictionary encode for better efficiency. I am curious what are you plans generally along these lines and if there are any opportunities for standardizing different string memory layouts (Weld may need to support more than one memory layout) to make it easier for other systems to integrate with Weld.

cc @julienledem

**Comment by deepakn94** [comment]
Yeah, this makes sense. I don't think we're wed to any particular data format at this point -- it would be nice if Weld uses the same data format Arrow uses.

One of the items we have on our roadmap is figuring out how to get Weld to operate on arbitrary input data formats (that is, you specify a Weld computation assuming a particular data format, but then figure out how to transform the computation to accommodate other data formats).

**Comment by wesm** [comment]
The notion of "arbitrary input data formats" is potentially a rathole. Beyond non-nullable tensor-like memory (i.e. the NumPy ndarray model), packed record / row-oriented tables (similar to Spark's Tungsten "off heap row" tuple object), and nullable nested columnar SQL-style memory (e.g. in the style of  Arrow, Drill, and Spark SQL's in-memory columnar), what other kinds of memory formats do you want to support? When you go beyond these types of memory models, you run into data structures where compile-time value access logic is difficult or impossible; then your compiler will have to generate a sort of "dynamic iterator" logic. 

It's probably reasonable to expect users to conform their data to one of your supported memory models, for which you know how to generate optimized IR.

One possibility is to define "Weld metadata" and a C API for third parties to implement to provide a mapping between their data format and the way that Weld wants to iterate over the data structures (similar to foreign data wrappers in Postgres and other SQL systems).

**Comment by jjthomas** [comment]
@wesm On a related note, one thing we were thinking about was generating parsers for JSON, CSV, or other textual formats that take advantage of our knowledge of the fields used by the query, predicates applied, etc. Do you think this will be useful for pandas workloads?
